File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/BlePresenceScanPort.kt
package expo.modules.vescapecore.connection

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.pm.PackageManager
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.os.Build
import android.util.Log
import expo.modules.vescapecore.service.VESC_SESSION_TAG

/**
 * Real Bluetooth behind [PresenceScanPort]. Readiness is reported the moment `startScan` is accepted
 * by the adapter, which is what makes the five-second Presence Scan window start on a usable radio
 * rather than at foreground entry.
 *
 * @parity /modules/vescape-core/ios/connection/BlePresenceScanPort.swift
 */
@SuppressLint("MissingPermission")
internal class BlePresenceScanPort(private val context: Context) : PresenceScanPort {
    private var callback: ScanCallback? = null

    private val adapter
        get() = (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter

    override fun bluetoothEnabled(): Boolean = adapter?.isEnabled == true

    override fun scanPermissionGranted(): Boolean =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            context.checkSelfPermission(Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED
        } else {
            context.checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        }

    override fun scannerAvailable(): Boolean = adapter?.bluetoothLeScanner != null

    override fun startScan(
        onReady: () -> Unit,
        onObserved: (bleId: String, rssi: Int?) -> Unit,
        onFailed: (message: String) -> Unit,
    ): Boolean {
        val scanner = adapter?.bluetoothLeScanner ?: return false
        val cb = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                onObserved(result.device.address, result.rssi)
            }

            override fun onScanFailed(errorCode: Int) {
                onFailed("BLE scan failed: $errorCode")
            }
        }
        return try {
            scanner.startScan(
                null,
                ScanSettings.Builder()
                    .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                    .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES)
                    .build(),
                cb,
            )
            callback = cb
            onReady()
            true
        } catch (e: Exception) {
            Log.w(VESC_SESSION_TAG, "Presence scan start failed: ${e.message}")
            false
        }
    }

    override fun stopScan() {
        val cb = callback ?: return
        callback = null
        try {
            adapter?.bluetoothLeScanner?.stopScan(cb)
        } catch (e: Exception) {
            Log.w(VESC_SESSION_TAG, "Presence scan stop failed: ${e.message}")
        }
    }
}

File: modules/vescape-core/ios/connection/BlePresenceScanPort.swift
import CoreBluetooth
import Foundation

/// Real CoreBluetooth behind `PresenceScanPort`.
///
/// The session central is shared with the Board Session, so this adapter does not own callbacks: it
/// starts and stops discovery and receives readiness / advertisements forwarded by
/// `BoardSessionController`'s `VescGattListener` conformance.
///
/// Readiness is the central reaching `.poweredOn` — the moment the five-second Presence Scan window
/// may start, which on iOS is meaningfully later than foreground entry.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/BlePresenceScanPort.kt
internal final class BlePresenceScanPort: PresenceScanPort {
  private let central: () -> CBManagerState
  private let start: () -> Void
  private let stop: () -> Void

  private var onReady: (() -> Void)?
  private var onObserved: ((String, Int?) -> Void)?
  private var onFailed: ((String) -> Void)?

  init(
    central: @escaping () -> CBManagerState,
    start: @escaping () -> Void,
    stop: @escaping () -> Void
  ) {
    self.central = central
    self.start = start
    self.stop = stop
  }

  func bluetoothEnabled() -> Bool {
    // `.unknown` / `.resetting` are "not yet answered", not "off": the scan waits for readiness
    // instead of being refused with `bluetooth_disabled`.
    central() != .poweredOff
  }

  func scanPermissionGranted() -> Bool {
    CBCentralManager.authorization == .allowedAlways || central() != .unauthorized
  }

  func scannerAvailable() -> Bool { central() != .unsupported }

  func startScan(
    onReady: @escaping () -> Void,
    onObserved: @escaping (String, Int?) -> Void,
    onFailed: @escaping (String) -> Void
  ) -> Bool {
    self.onReady = onReady
    self.onObserved = onObserved
    self.onFailed = onFailed
    start()
    // A central already powered on never emits another state change, so readiness is reported here.
    if central() == .poweredOn { onReady() }
    return true
  }

  func stopScan() {
    onReady = nil
    onObserved = nil
    onFailed = nil
    stop()
  }

  // MARK: - Forwarded from the shared `VescGattListener`

  /// The central reached `.poweredOn` after the scan was requested.
  func deliverReady() {
    onReady?()
  }

  func deliverObserved(id: String, rssi: Int) {
    onObserved?(id, rssi)
  }

  func deliverFailure(_ message: String) {
    onFailed?(message)
  }
}
